### ISSUE TITLE
Loop gets stuck in criterion function :S

### ISSUE THREAD

**Issue by shwe87** [pytorch/xla#2229]
Hello,

I am totally new to using TPUs and I followed these steps:
!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py --version nightly --apt-packages libomp5 libopenblas-dev

import torch_xla
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met
import torch_xla.distributed.parallel_loader as pl
import torch_xla.distributed.xla_multiprocessing as xmp
import torch_xla.utils.utils as xu

In my training loop, I have this where it gets stuck, in the criterion bit
            model.zero_grad()
            # get the output from the model
            print("Call model")
            

            output, h = model(specs, h)
            output = F.log_softmax(output, dim=2)
            output = output.transpose(0,1)
            # calculate the loss and perform backprop
           print("Softmax done")
            loss = criterion(output, labels.float(), input_lengths, label_lengths)
            print("Criterion done")
            loss.backward()

            # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
            nn.utils.clip_grad_norm_(model.parameters(), clip)
            xm.optimizer_step(optimizer)
            scheduler.step()

**My optimizer, scheduler and criterion:**
optimizer = torch.optim.Adam(my_rnn_net.parameters(), lr=lr)
scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, max_lr=lr, 
                                            steps_per_epoch=int(len(train_loader)),
                                            epochs=epochs,
                                            anneal_strategy='linear')

criterion = nn.CTCLoss(blank=28, zero_infinity=True)

Never prints criterion, only until Softmax done.

With GPU it works fine.
I am really confused, thanks.

**Comment by dlibenzi** [comment]
Hi!
Do you mind providing a working (or better ... not working in this case 😄 ) Colab?

**Comment by shwe87** [comment]
Hello,
Here it is:

https://colab.research.google.com/drive/1WQ0g_9w08olOS05dBPfmVVafjBiQwBPc

Thanks so much!

**Comment by dlibenzi** [comment]
Can you make it public to all the ones with the link?
(can't access)

**Comment by shwe87** [comment]
Sorry, how can I make the notebook public in Google Colab? :S

**Comment by dlibenzi** [comment]
The code is pretty big and requires data loaded from your drive.
This is not part of your problem but you need to use XLA serialization to serialize XLA tenors:

https://github.com/pytorch/xla/blob/master/API_GUIDE.md#saving-and-loading-xla-tensors

For Colab (low memory) it is better to use:
```Python
import torch_xla.utils.serialization as xser

xser.save(model.state_dict(), path)
...
state_dict = xser.load(path)
model.load_state_dict(state_dict)
```

About your loss not being printed, your model is using CTC loss, which we do not lower ATM.
It is possible that this ends up creating a lot of CPU<->TPU transitions which slows down the model a lot, inside the loss function.
How long did you wait before saying that it does not print?

**Comment by shwe87** [comment]
- With GPU, it takes like less than a second to run the loss function.
- With kaggle TPU and colab TPU, it was taking too long, may be 10-12 minutes and it didn't print at all, so I just interrupted it.

**Comment by dlibenzi** [comment]
GPUs works different from XLA.
GPUs dispatch one operation at a time, while XLA combines all the possible operations inside one graph, compiles it, and executes it.
If the graph does not change across steps, the compilations are cached, so the following steps will never have to wait anymore.
10..12 minutes is a lot though.

Any chance the Colab can be modified to not require access to your Google Drive?

**Comment by shwe87** [comment]
Actually, I am using dataset from kaggle. I am using drive for saving checkpoints. This box is commented in the code to avoid duplications.

! kaggle datasets download -d bryanpark/spanish-single-speaker-speech-dataset

**Comment by shwe87** [comment]
> GPUs works different from XLA.
> GPUs dispatch one operation at a time, while XLA combines all the possible operations inside one graph, compiles it, and executes it.
> If the graph does not change across steps, the compilations are cached, so the following steps will never have to wait anymore.
> 10..12 minutes is a lot though.
> 
> Any chance the Colab can be modified to not require access to your Google Drive?

Thank you so much. I have limited resources and I am using all the resources of GPU available in kaggle and Google Colab. Would also like to use TPU, that's why I am trying to run my codes in TPU..... but without success....
